pkg/api/validate/dynamic/serviceprincipal_test.go

```diff
@@ -0,0 +1,127 @@
+package dynamic
+
+// Copyright (c) Microsoft Corporation.
+// Licensed under the Apache License 2.0.
+
+import (
+	"context"
+	"crypto/hmac"
+	"crypto/sha512"
+	"encoding/base64"
+	"encoding/json"
+	"net/url"
+	"testing"
+
+	"github.com/Azure/go-autorest/autorest/adal"
+	"github.com/golang/mock/gomock"
+	"github.com/sirupsen/logrus"
+
+	"github.com/Azure/ARO-RP/pkg/util/azureclient"
+	mock_aad "github.com/Azure/ARO-RP/pkg/util/mocks/aad"
+)
+
+func TestValidateServicePrincipal(t *testing.T) {
```

Comment:
It might also be possible to test for a failure to get the token, and a failure to parse the token (lines 21 and 28 of serviceprincipal.go)

Reply:
That should be possible by creating multiple expects and returning those errors. I didn't create those initially because it's a mocked return, so it should hit the `if err != nil { return err }` line.
I can add some tests to cause the parser to fail as well.